### PR TITLE
Create Taproot feature and introduce signer type enum

### PR DESCRIPTION
**Pull request by arik-so**

### CODE REVIEW

File: lightning/src/sign/taproot.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	/// sent to us and checking the HTLCs.
+	///
+	/// The preimages of outgoing HTLCs that were fulfilled since the last commitment are provided.
+	/// A validating signer should ensure that an HTLC output is removed only when the matching
+	/// preimage is provided, or when the value to holder is restored.
+	///
+	/// Note that all the relevant preimages will be provided, but there may also be additional
+	/// irrelevant or duplicate preimages.
+	//
+	// TODO: Document the things someone using this interface should enforce before signing.
```
Comment (ariard) [inline lightning/src/sign/taproot.rs:29]: The signer should verify we’re spending a P2TR.

File: lightning/src/sign/taproot.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	fn partially_sign_counterparty_commitment(&self, commitment_tx: &CommitmentTransaction,
+		preimages: Vec<PaymentPreimage>, secp_ctx: &Secp256k1<secp256k1::All>,
+	) -> Result<(PartialSignature, Vec<Signature>), ()>;
```
Comment (ariard) [inline lightning/src/sign/taproot.rs:32]: Do we expect this method to return `Signature` for each, if yes also the `inclusion_proof` for the script path as we expect this method to be use internally by LDK for on-chain claims?

File: lightning/src/sign/taproot.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+	/// Create a signature for the given input in a transaction spending an HTLC transaction output
+	/// or a commitment transaction `to_local` output when our counterparty broadcasts an old state.
+	///
+	/// A justice transaction may claim multiple outputs at the same time if timelocks are
+	/// similar, but only a signature for the input at index `input` should be signed for here.
+	/// It may be called multiple times for same output(s) if a fee-bump is needed with regards
+	/// to an upcoming timelock expiration.
+	///
+	/// Amount is value of the output spent by this input, committed to in the BIP 143 signature.
```
Comment (ariard) [inline lightning/src/sign/taproot.rs:63]: BIP341 signature if we’re talking about PT2R outputs no ?

File: lightning/src/ln/features.rs [hunk 4]
```diff
@@ -384,6 +384,9 @@ mod sealed {
 	define_feature!(27, ShutdownAnySegwit, [InitContext, NodeContext],
 		"Feature flags for `opt_shutdown_anysegwit`.", set_shutdown_any_segwit_optional,
 		set_shutdown_any_segwit_required, supports_shutdown_anysegwit, requires_shutdown_anysegwit);
+	define_feature!(31, Taproot, [InitContext, NodeContext, ChannelTypeContext],
```
Comment (dunxen) [inline lightning/src/ln/features.rs:387]: Should we add a `TODO(taproot)` to the module docs above to remember to document support for simple taproot channels when the time comes?

**Pull request closed without merging**